library/std/src/primitive_docs.rs

```diff
@@ -194,14 +194,47 @@ mod prim_bool {}
 /// # `!` and traits
 ///
 /// When writing your own traits, `!` should have an `impl` whenever there is an obvious `impl`
-/// which doesn't `panic!`. As it turns out, most traits can have an `impl` for `!`. Take [`Debug`]
+/// which doesn't `panic!`. The reason is that functions returning an `impl Trait` cannot have
+/// divergence, i.e., returning `!`, as their only possible code path. As an example, this code
+/// doesn't compile:
+///
+/// ```compile_fail
+/// use core::ops::Add;
+///
+/// fn foo() -> impl Add<u32> {
+///     unimplemented!()
+/// }
+/// ```
+///
+/// While this code does:
+///
+/// ```
+/// use core::ops::Add;
+///
+/// fn foo() -> impl Add<u32> {
+///     if true {
+///         unimplemented!()
+///     } else {
+///         0
+///     }
+/// }
+/// ```
+///
+/// The reason is that, in the first example, there are many possible types for `!` to coerce
+/// to, because the function's return value is polymorphic. However, in the second example, the
```

Comment:
I don't know what polymorphic means in this context. I thought that meant like a type constructor that can be instantiated with multiple different types.

Reply:
My understanding is that polymorphic just means "can be one of many concrete types".

Reply:
I could be wrong though.

Reply:
Hmm I'm not sure one way or another, but either way I think it would be more clear to say 'multiple possible return types'.

Reply:
How about:

```suggestion
/// to, because the function can return one of many concrete types. However, in the second example, the
```